rclpy/rclpy/action/client.py

```diff
@@ -177,6 +176,11 @@ def __init__(
         self._result_sequence_number_to_goal_id = {}
         # key: UUID in bytes, value: callback function
         self._feedback_callbacks = {}
+        self._data_lock = threading.Lock()
```

Comment:
I would suggest we can introduce the threading lock against the specific member data object, instead of having giant lock in the `ActionClient`. that can be also more performative.

see https://github.com/ros2/rclcpp/blob/e2965831d51e9be03470cb07f8721012afcade9b/rclcpp_action/src/client.cpp#L112-L119